### PR TITLE
Add first set of tools to generate the electorate

### PR DESCRIPTION
**Pull request by tbreeds**
Signed-off-by: Tony Breeds <tony@bakeyournoodle.com>

### CODE REVIEW

File: elections/tools/README.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+
+# Election Tools
+
+This directory contains the tools required to generate the kata electorate
+
+## generate_electorate.py
+
+This tool need the following python libraries pytz github3.py pyyaml.  To
```
Comment (grahamwhaley) [inline elections/tools/README.md]: s/need/needs/

Maybe make the requirements a bullet list:

- pytz
- github3.py
- pyyaml

File: elections/tools/README.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+install them in a virtutal envoronment do something like:
+
+```
+python3 -m venv .venv
```
Comment (grahamwhaley) [inline elections/tools/README.md]: We normally prefix our code commands with `$ ` - have a peek at https://github.com/kata-containers/documentation/blob/master/Documentation-Requirements.md which I think has some documentation style notes.

File: elections/tools/README.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+Before running the tool you will need to create a
+[github API token](https://github.blog/2013-05-16-personal-api-tokens/)
+
+replace ``__API_TOKEN__`` with your personal token
+
+Also update the election start and end times.  Then run the tool with:
+``.venv/bin/python ./generate_electorate.py``
+
+The code looks a all commits in all kata-containers repos *except*
```
Comment (grahamwhaley) [inline elections/tools/README.md]: s/a/at/

File: elections/tools/README.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+kata-containers/linux and kata-containers/qemu.  As both of these are forks (in
+the GitHUb sense) they'll have lots of contributors that may not be kata
+contributors.
+
+For contributors that have more than one email address I pick one as default
```
Comment (grahamwhaley) [inline elections/tools/README.md]: s/I/it/

File: elections/tools/README.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+but supply all the others so you can be smarter about emailing them it you
```
Comment (grahamwhaley) [inline elections/tools/README.md]: s/supply/supplies/
s/you can..../so we can be smarter about where to send the emails/

File: elections/tools/generate_electorate.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
+#!/usr/bin/env python3
+
```
Comment (grahamwhaley) [inline elections/tools/generate_electorate.py]: You probably want to add an [SPDX license header](https://github.com/kata-containers/documentation/blob/master/Licensing-strategy.md) in here - the rest of the project uses SPDX style and is predominantly apache2.

File: elections/tools/generate_electorate.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+end_time = datetime.datetime(2018, 8, 1, 0, 0, 0, tzinfo=pytz.UTC)
+# ... Or run just include all commits
+# start_time = end_time = None
+# All commits
+number = -1
+projects = []
+
+author_cache = {}
+for repo in org.repositories():
+    # These repos have been around since before Kata containers and I'm
```
Comment (grahamwhaley) [inline elections/tools/generate_electorate.py]: Correct. Let's update that comment maybe to something like:
```
Skip these repos are they are not a core part of the project, and are forked/imported so container many contributors from outside the project.
```

File: elections/tools/README.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+[github API token](https://github.blog/2013-05-16-personal-api-tokens/)
+
+replace ``__API_TOKEN__`` with your personal token
+
+Also update the election start and end times.  Then run the tool with:
+``.venv/bin/python ./generate_electorate.py``
+
+The code looks a all commits in all kata-containers repos *except*
+kata-containers/linux and kata-containers/qemu.  As both of these are forks (in
+the GitHUb sense) they'll have lots of contributors that may not be kata
```
Comment (jodh-intel) [inline elections/tools/README.md]: Nit: "GitHub".

File: elections/tools/README.md [hunk 9]
```diff
@@ -0,0 +1,37 @@
+
+# Election Tools
+
+This directory contains the tools required to generate the kata electorate
+
+## generate_electorate.py
+
+This tool need the following python libraries pytz github3.py pyyaml.  To
+install them in a virtutal envoronment do something like:
```
Comment (jodh-intel) [inline elections/tools/README.md]: Typos: "virtual environment".

File: elections/tools/README.md [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+```
+python3 -m venv .venv
+.venv/bin/pip install pytz github3.py pyyaml
+```
+
+Before running the tool you will need to create a
+[github API token](https://github.blog/2013-05-16-personal-api-tokens/)
```
Comment (jodh-intel) [inline elections/tools/README.md]: Nit: Missing period at end of sentence.

File: elections/tools/README.md [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+replace ``__API_TOKEN__`` with your personal token
```
Comment (jodh-intel) [inline elections/tools/README.md]: > Replace `__API_TOKEN__` with your personal token.

You could use `sed -i` here, but this is fine and I suspect you've specified it like this to maximise security on a multi-user system (by avoiding `ps` seeing the token value).

File: elections/tools/README.md [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+the GitHUb sense) they'll have lots of contributors that may not be kata
+contributors.
+
+For contributors that have more than one email address I pick one as default
+but supply all the others so you can be smarter about emailing them it you
+want.
+
+The sources for email addresses are:
+- github account
+- git commit data
```
Comment (jodh-intel) [inline elections/tools/README.md]: Nit: "Git" as this is referring to the name of the tool, not the command itself.

File: elections/tools/README.md [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+kata-containers/linux and kata-containers/qemu.  As both of these are forks (in
+the GitHUb sense) they'll have lots of contributors that may not be kata
+contributors.
+
+For contributors that have more than one email address I pick one as default
+but supply all the others so you can be smarter about emailing them it you
+want.
+
+The sources for email addresses are:
+- github account
```
Comment (jodh-intel) [inline elections/tools/README.md]: Nit: "GitHub" (here and elsewhere).

File: elections/tools/generate_electorate.py [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+            if ((author.email is None or
+                    'users.noreply.github.com' in author.email) and
+                    match.group('email')):
+                author.email = match.group('email')
+            if author.name is None and match.group('name'):
+                author.name = match.group('name')
+        authors.add(author)
+    projects.append({str(repo): authors})
+
+# Dark YAML voodoo
```
Comment (jodh-intel) [inline elections/tools/generate_electorate.py]: Love it :smile:

**Comment by raravena80** [comment]
@tbreeds any updates on this? Thx

**Comment by tbreeds** [comment]
> @tbreeds any updates on this? Thx

I'm sorry I'm not sure what's required of me.  I feel like I've addressed the initial reviews by @jodh-intel and @grahamwhaley I can see it needs a rebase / merge update but what else is required?

**Comment by tbreeds** [comment]
Thanks all!

**Pull request merged**